docs: Adding more docs on useful commands in readme (#689)
edx/devstack: docs/devstack_interface.rst

Devstack Interface
------------------

Devstack comes built in with many useful make commands that act as an user interface. This UI is essentially a make wrapper around docker-compose commands. We attempt to give a short summary of what the make commands do below, but it would be a good idea for you to familiarize yourself with some knowledge of docker-compose.

> Due to the organic nature of how this user interface came into being, there are often multiple ways to do the same thing. The two main variants of commands are of the form: ``dev.ACTION.SERVICE`` vs ``SERVICE-ACTION``. The ``SERVICE-ACTION`` variant tends to be more automatic command-completion friendly.

Reviewer: ```suggestion
Due to the organic nature of how this user interface came into being, there are often multiple ways to do the same thing. The two main variants of commands are of the form: ``dev.ACTION.SERVICE`` vs ``SERVICE-ACTION``. The ``SERVICE-ACTION`` variant tends to be more tab-completion friendly.
```
Suggested rephrasing -- I think I'd understand it better this way, personally.

Examples:

.. code:: sh

    make dev.up.registrar
    make registrar-up

    make dev.shell.lms
    make lms-shell

    make dev.logs.gradebook
    make gradebook-logs

The user interface for devstack often also gives you both big hammers(``make dev.pull``) and small hammers(``make dev.pull.<service>``) to do things. It is recommend you opt for the small hammer commands, because they often tend to be alot faster.

Useful Commands and Summary
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. Note: this document does not contain all commands in Makefile. To see full range of the make interface, please see Makefile

- ``dev.pull.<service>`` - Pull latest Docker images for the service and its dependencies

>   When to use: When you only want to update images for a subset of devstack. Such as when you are only working on lms, use ``make dev.pull.lms`` to pull images for lms and it dependencies

Reviewer: It might be useful to suggest scenarios here, or explain _why_ someone might want new images. ...I don't actually know the full answer to that, beyond getting latest Python reqs, but it might look something like this:
```suggestion
  When to use: When you want to start working with a service (testing out changes, running unit tests, etc.) and want the latest built image since it has been a few days since you last worked with it. Particularly useful when the Python requirements files have changed recently.
```

Author: I like this suggestion!

  Variations:

  + ``make dev.pull`` or ``make pull`` will pull all images in the default devstack.

    When to use: Probably only when you are first setting up devstack. Do not use this often. This will take a lot of time.

  + ``make dev.pull.<service1>+<service2>`` will pull images for <service1>, <service2>, and their dependencies

  + ``make dev.pull.without-deps.<service>`` will only pull <service> image.

    When to use: If you only want to update one image and do not mind if the other images are behind latest.

- ``dev.up.<service>`` - Create and start containers. i.e. brings up the <service> container and its dependencies

>   When you are working on a specific service, it is recommended you use this command to bring up the necessary containers for your service i.e if working in lms, use ``make dev.up.lms`` to bring up containers for lms and its dependencies.

Reviewer: Maybe mention ordering here -- pull first, if desired, then up. Not sure if it helps, though.

Reviewer: Likely worthwhile to mention the counterparts of this command, something like: "See also `stop` and `down`."

Author: Added a little snippet for this

  Variations:

>   + ``make dev.up`` or ``make up`` will bring up containers for everything in default devstack

Reviewer: Emphasizing the "everything" may help warn people away from this:
```suggestion
  + ``make dev.up`` or ``make up`` will bring up containers for *everything* in default devstack
```

>   + ``make dev.up.<service1>+<service2>`` will pull images for <service1>, <service2>, and their dependencies

Reviewer: ```suggestion
  + ``make dev.up.<service1>+<service2>`` will bring up <service1>, <service2>, and their dependencies
```

  + ``make dev.up.without-deps.<service>`` will only bring up the <service> container

- ``dev.stop.<container>``: only stops the container. This does not remove the container or the networks it has created

  When to use: When you are pausing your work on this container/devstack and you want to pick back up from where you left off. Next time you use dev.up to bring up containers, you should be able to mostly pick back up from where you started.  TODO(reviewer): Is this true in your experience.

> - dev.down.<container>: stops the specified container and also removes the stopped containers as well as any networks that were created. Next time you use dev.up to bring up container, your container have reverted back to the pulled image.  This will not affect content of the databases.

Reviewer: Just some missing code-quotes:

```suggestion
- ``dev.down.<container>``: stops the specified container and also removes the stopped containers as well as any networks that were created. Next time you use dev.up to bring up container, your container have reverted back to the pulled image.  This will not affect content of the databases.
```

  When to use: use this command only if you are okay with removing any changes you might have made to the container

  Variation: ``make dev.down`` will stop all your containers

- ``dev.shell.<container>``: used to enter the shell of the specified container.

  When to use: To update python packages, to run migrations, or any shell commands you want to run on the running service

  Variation: ``make <service>-shell``

- ``dev.attach.<container>``: dev.up is setup to bring up the container/service in the background. This attaches the container to your shell.

  When to use: If you put a breakpoint somewhere in your code, the pdb shell will show up in here. TODO(reviewer): help with words to explain this use case.

>   Variation: ``make <service>-shell``

Reviewer: ```suggestion
  Variation: ``make <service>-attach``
```

> - ``dev.logs.<container>``: View the logs of the specified service <container>

Reviewer: Some of these commands use `<service>` and others use `<container>`. Is there a difference?

Author: I've been using service for cases where the command will affect more than just the <service> container. So differentiating between only affect lms vs lms and its dependencies

Author: This seemed like it might help, but there is a chance this is just confusing

Reviewer 2: I recomend sticking with `<service>`, since services and containers are different things in docker-compose, and (annoyingly) are named different sometimes.

Example: the service `edx_notes_api` spins up a container named `edxnotesapi`. The devstack commands expect `dev.*.edx_notes_api`, not `dev.*.edxnotesapi`.

Reviewer: If I'm understanding this correctly, it might be best to draw the distinction by saying "service" and "service plus dependencies" in the prose.

Author: Yep, I try to say "service and its dependencies" whereever that distinction is key.

  When to use: during debugging, this would help you see live logs coming out of your container.

  Variation: ``make dev.logs`` to view logs for all running containers. Do not use this! This is likely very overwhelming.

  Variation: ``make <container>-logs``

- ``dev.restart-container.<container>`` restarts container. TODO(jinder): change any references to this in documentation
  Note: this will only restart <container> and not its dependencies

>   When to use: TODO(reviewer): help

Reviewer: I think this is basically the bigger-hammer version of restart-devserver, just based on the old docs for it.

Reviewer 2: Without getting into the details, yes.

Due to the organic nature of how this user interface came into being, there are often multiple ways to do the same thing. The two main variants of commands are of the form: ``dev.ACTION.SERVICE`` vs ``SERVICE-ACTION``. The ``SERVICE-ACTION`` variant tends to be more tab-completion friendly.

Examples:

.. code:: sh

    make dev.up.registrar
    make registrar-up

    make dev.shell.lms
    make lms-shell

    make dev.logs.gradebook
    make gradebook-logs

The user interface for devstack often also gives you both big hammers(``make dev.pull``) and small hammers(``make dev.pull.<service>``) to do things. It is recommend you opt for the small hammer commands, because they often tend to be alot faster.

Useful Commands and Summary
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. Note: this document does not contain all commands in Makefile. To see full range of the make interface, please see Makefile

- ``dev.pull.<service>`` - Pull latest Docker images for the service and its dependencies

>   When to use: When you want to start working with a service (testing out changes, running unit tests, etc.) and want the latest built image since it has been a few days since you last worked with it. Particularly useful when the Python requirements files have changed recently.

Reviewer 2: I think a big point of confusion is "what do new images actually contain? / what do I get by pulling them?". In the current iteration of devstack, the answer is essentially - requirements. You get an environment with a fresh set of requirements, and it persists. This is in contast to `make requirements` , which only persists for as long as the service is up.

Author: thanks for this clarification

  Variations:

  + ``make dev.pull`` or ``make pull`` will pull all images in the default devstack.

    When to use: Probably only when you are first setting up devstack. Do not use this often. This will take a lot of time.

  + ``make dev.pull.<service1>+<service2>`` will pull images for <service1>, <service2>, and their dependencies

  + ``make dev.pull.without-deps.<service>`` will only pull <service> image.

>     When to use: If you only want to update one image and do not mind if the other images are behind latest.

Reviewer 2: May be worth noting that the service must be brought completely down and then back up for the new image to take effect.

Reviewer: Yes, interactions like this seem really important to document.

Author: Added a Note on this on line 34.

  When to use: When you are pausing your work on this container/devstack and you want to pick back up from where you left off. Next time you use dev.up to bring up containers, you should be able to mostly pick back up from where you started.  TODO(reviewer): Is this true in your experience.

- ``dev.down.<container>``: stops the specified container and also removes the stopped containers as well as any networks that were created. Next time you use dev.up to bring up container, your container have reverted back to the pulled image.  This will not affect content of the databases.

  When to use: use this command only if you are okay with removing any changes you might have made to the container

  Variation: ``make dev.down`` will stop all your containers

- ``dev.shell.<container>``: used to enter the shell of the specified container.

  When to use: To update python packages, to run migrations, or any shell commands you want to run on the running service

  Variation: ``make <service>-shell``

- ``dev.attach.<container>``: dev.up is setup to bring up the container/service in the background. This attaches the container to your shell.

>   When to use: If you put a breakpoint somewhere in your code, the pdb shell will show up in here. TODO(reviewer): help with words to explain this use case.

Reviewer 2: I suggest mentioning:
* you can use `Ctrl+c` to restart the devserver, when attached
* you can use `Ctrl+p Ctrl+q` to detach without closing your terminal

Author: These are good tips, thanks!

- ``dev.pull.<service>`` - Pull latest Docker images for the service and its dependencies

  When to use: If you have not used Devstack for a while or pulled new images for a while, the installed requirements in your containers might no longer match those used by your code. So it is recommended you pull new images whenever you want your containers to have the latest requirements(python libraries...) installed.

>   Note: for new service images to be used, you first need to bring down those services and then bring them back after a pull.

Reviewer: ```suggestion
  Note: for new service images to be used, you first need to bring down those services and then bring them back up after a pull.
```
Suggested wording change, to hint at the command people would use.
